flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -1329,33 +1364,66 @@ public void tryHandleCheckpointException(CheckpointMetaData checkpointMetaData,
 	 */
 	public final class ActionContext {
 
-		private final Runnable actionUnavailableLetter = ThrowingRunnable.unchecked(mailbox::waitUntilHasMail);
+		private final Runnable actionUnavailableLetter;
+		private final TaskMailboxExecutorServiceImpl mailboxExecutorService;
+
+		public ActionContext(TaskMailboxExecutorServiceImpl mailboxExecutorService) {
+			final Mailbox mailbox = taskMailboxExecutor.getMailbox();
+			this.actionUnavailableLetter = ThrowingRunnable.unchecked(() -> mailbox.takeMail().run());
+			this.mailboxExecutorService = mailboxExecutorService;
```

Comment:
Nitpick: you are mixing two fields in the `ActionContext` implementation `TaskMailboxExecutorServiceImpl 
 mailboxExecutorService // this` and `TaskMailboxExecutorServiceImpl taskMailboxExecutor // enclosing StreamTask`.

Although, I'd expect that they both reference the same object.